### PR TITLE
ARROW-11089: [C++][Gandiva] Support list datatype for gandiva UDF 

### PR DESCRIPTION
**Pull request by NinaPeng**
Hi there, this PR is going to support list type for gandiva expression inputs and outputs.
- add `list<utf8>`, `list<int>`, `list<float>` type for expression inputs and outputs
- add two function for `list<utf8>` inputs: `array_length(list<utf8>)`, `array_contains(list<utf8>, utf8)`

TODO:
- current structure can support `list<decimal>` type, but should handle extra precision and scale
- as for `list<bool>` type, should add extra function to get bits list for computations

This is my first PR, I am unfamiliar with workflow for contributing code. If there is anything I was missing, please kindly tell me to update

### CODE REVIEW

File: cpp/src/gandiva/annotator.cc [hunk 1]
```diff
@@ -46,15 +46,23 @@ FieldDescriptorPtr Annotator::MakeDesc(FieldPtr field, bool is_output) {
   int data_idx = buffer_count_++;
   int validity_idx = buffer_count_++;
   int offsets_idx = FieldDescriptor::kInvalidIdx;
+  int child_offsets_idx = FieldDescriptor::kInvalidIdx;
```
Comment (NinaPeng) [inline cpp/src/gandiva/annotator.cc:49]: as for list type array data:
- list<binary-like> array data structure:
image
- list<primitive> array data structure:
image

**Comment by NinaPeng** [comment]
Currently, Java JNI checks was failed.

It is because that jni ExtGandivaType is not support List type

**Comment by emkornfield** [comment]
@NinaPeng I think we would need the JNI tests fixed before merging.  @praveenbingo or @pravindra would you have time to review?

**Comment by NinaPeng** [comment]
@emkornfield Thanks for your attention. But I have no idea how to add list type for jni ExtGandivaType. Would you like to give some suggestions?

**Comment by emkornfield** [comment]
> . But I have no idea how to add list type for jni ExtGandivaType.

I'm sorry, I'm very familiar with this code.

**Comment by projjal** [comment]
> @emkornfield Thanks for your attention. But I have no idea how to add list type for jni ExtGandivaType. Would you like to give some suggestions?

You need to be able to convert convert the information in ExtGandivaType to convert into an arrow list type. This happens in gandiva/jni/jni_common.cc

**Comment by projjal** [comment]
> > @emkornfield Thanks for your attention. But I have no idea how to add list type for jni ExtGandivaType. Would you like to give some suggestions?
> 
> You need to be able to convert convert the information in ExtGandivaType to convert into an arrow list type. This happens in gandiva/jni/jni_common.cc

And the corresponding arrow type to proto is done in jni/expression_registry_helper.cc. This is where it is likely causing issue when exporting the registered functions (which include new functions with list type)

File: cpp/src/gandiva/field_descriptor.h [hunk 2]
```diff
@@ -58,12 +63,15 @@ class FieldDescriptor {
...
 
+  bool HasChildOffsetsIdx() const { return child_offsets_idx_ != kInvalidIdx; }
+
  private:
   FieldPtr field_;
   int data_idx_;
   int validity_idx_;
   int offsets_idx_;
   int data_buffer_ptr_idx_;
+  int child_offsets_idx_;
```
[the remark this answers is not in the document]
Comment (NinaPeng) [reply cpp/src/gandiva/field_descriptor.h:74]: Got it. It's better to add child validity index for more list type computation. I'll fix it as soon as possible.

File: cpp/src/gandiva/projector.cc [hunk 3]
```diff
@@ -301,7 +348,24 @@ Status Projector::AllocArrayData(const DataTypePtr& type, int64_t num_records,
   }
   buffers.push_back(std::move(data_buffer));
 
-  *array_data = arrow::ArrayData::Make(type, num_records, std::move(buffers));
+  if (type->id() == arrow::Type::LIST) {
+    auto internal_type = type->field(0)->type();
+    ArrayDataPtr child_data;
+    if (arrow::is_primitive(internal_type->id())) {
+      child_data = arrow::ArrayData::Make(internal_type, 0 /*initialize length*/,
+                                          {nullptr, std::move(buffers[2])}, 0);
```
[the remark this answers is not in the document]
Comment (NinaPeng) [reply cpp/src/gandiva/projector.cc:356]: At the begining, I just want to support some simple list type computation. So I make an assumption without validity information. I also ignored child validity in field descriptor. 

Will think about child validity addtionally.

**Comment by amol-** [comment]
Closing because it has been untouched for a while, in case it's still relevant feel free to reopen and move it forward 👍

File: cpp/src/gandiva/field_descriptor.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (lriggs) [reply cpp/src/gandiva/field_descriptor.h:74]: @NinaPeng I'm working on resurrecting this PR to add suport for lists in Gandiva. I got things working through java/jni but am now trying to add the validity vectors.  Do you have any more details on the approach mentioned above? Or maybe a PR I could look at? Thanks.

**Pull request closed without merging**